### ISSUE TITLE
'OverflowError' when calling 'cupy.einsum' for batched GEMM (even with quite small tensors)

### ISSUE THREAD

**Issue by mscipio** [cupy/cupy#1275]
Hi everybody!
I started experimenting a little bit with CuPy einsum functionality and, at my very first attempt, I found this weird error message.

```
OverflowError                             Traceback (most recent call last)
OverflowError: value too large to convert to int
Exception OverflowError: 'value too large to convert to int' in 'cupy.core.core._get_stride_for_strided_batched_gemm' ignored
```
I hope the screenshot attached below is clear enough to let you understand what I was trying.

It cannot be a problem with the memory of my GPU (Titan XP w/ 12GB).
As you can see the batched GEMM run using skcuda is totally fine, and it runs smoothly even if I change the first dimension of the two tensors to, let's say, 1.000.000, while CuPy is able to run only if the size is below, [100,24,5].

Any idea?!

![image](https://user-images.githubusercontent.com/17526160/40555918-f692a2c6-604a-11e8-8f4e-7238fed8308a.png)

**Comment by mscipio** [comment]
This is the error message I get on the first run (the previous one was probably a short version that appears once the GPU is already "messed up"):

```
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
OverflowError: value too large to convert to int
Exception OverflowError: 'value too large to convert to int' in 'cupy.core.core._get_stride_for_strided_batched_gemm' ignored
---------------------------------------------------------------------------
CUDADriverError                           Traceback (most recent call last)
<ipython-input-4-3a9bcc2065a1> in <module>()
      2 D_cp = cp.asarray(D)
      3 
----> 4 G_cp = cp.einsum('nij,ni->nj', J_cp, D_cp)
      5 print(G_cp.get().shape)
      6 np.allclose(G_res, G_cp.get())

/media/DATA/anaconda2/envs/occiput27/lib/python2.7/site-packages/cupy/linalg/einsum.pyc in einsum(*operands)
    345         result, subscript = single_views[0]
    346 
--> 347     result, subscript = calc_summed_view(result, subscript, output_subscript)
    348     result = calc_transposed_view(result, subscript, output_subscript)
    349     return result

/media/DATA/anaconda2/envs/occiput27/lib/python2.7/site-packages/cupy/linalg/einsum.pyc in calc_summed_view(ioperand, input_subscript, output_subscript)
     93 
     94     if axes_to_summed:
---> 95         result = ioperand.sum(axis=tuple(axes_to_summed)). \
     96             astype(ioperand.dtype)
     97     else:

cupy/core/core.pyx in cupy.core.core.ndarray.sum()

cupy/core/core.pyx in cupy.core.core.ndarray.sum()

cupy/core/reduction.pxi in cupy.core.core.simple_reduction_function.__call__()

cupy/util.pyx in cupy.util.memoize.decorator.ret()

cupy/core/reduction.pxi in cupy.core.core._get_simple_reduction_function()

cupy/core/reduction.pxi in cupy.core.core._get_simple_reduction_kernel()

cupy/core/carray.pxi in cupy.core.core.compile_with_cache()

/media/DATA/anaconda2/envs/occiput27/lib/python2.7/site-packages/cupy/cuda/compiler.pyc in compile_with_cache(source, options, arch, cache_dir, extra_source)
    159             cubin_hash = six.b(hashlib.md5(cubin).hexdigest())
    160             if hash == cubin_hash:
--> 161                 mod.load(cubin)
    162                 return mod
    163 

cupy/cuda/function.pyx in cupy.cuda.function.Module.load()

cupy/cuda/function.pyx in cupy.cuda.function.Module.load()

cupy/cuda/driver.pyx in cupy.cuda.driver.moduleLoadData()

cupy/cuda/driver.pyx in cupy.cuda.driver.check_status()

CUDADriverError: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered
```

**Comment by kmaehashi** [comment]
Thanks for the feedback.
Currently `cupy.einsum` uses unexpectedly large memory for intermediate variables, and the issue is going to be fixed in https://github.com/cupy/cupy/pull/1218.

**Comment by mscipio** [comment]
Thanks for the quick feedback!
Has that fix already been merged somewhere?

Do you suggest me to build a source version of CuPy fr GitHub? I was using the PyPi version 4.1.0 with cuda8.0 while getting the error ...

**Comment by kmaehashi** [comment]
Currently it's not merged (not included in v4.1.0 release). You can try building CuPy source from toslunar's `einsum` branch.

**Comment by kmaehashi** [comment]
@mscipio #1218 has been merged and will be included in the next release.
